### DISCUSSION TITLE
Make `context` optional in operators

### DISCUSSION THREAD

**Discussion by dstandish** [apache/airflow#13492] [Ideas]
Many operators don't use it.

And when developing locally, if your operator does not use it, it's very convenient to call `task.execute()` to test your task.

So that type would be `Union[Optional[dict], dict]` in essence.

WDYT?

**Comment by turbaszek** [comment]
I'm not convinced. All our tests do `op.execute(context={})` or similar and I don't see this as a something troublesome. Also, testing operators out side of the Airflow context is not best practice in my opinion. Hooks - yes, operators - no. 

But that's my subjective feeling 😄

**Reply by dstandish** [reply]
well my fingers say it _is_ troublesome ;)

**Reply by dstandish** [reply]
but... on serious note... given that it is _de facto_ optional, why not make it _explicitly_ optional?

**Comment by potiuk** [comment]
I'd say far better solution would be to have empty dict as default if not https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python . 

I think The problem with setting it as Optional would mean that the mutlitude of operators (the custom ones) would have to be prepared to accept 'execute' without context, which is not backwards-incompatible I am afraid. Of course we can make sure that Airflow itself does not do it and always passes a context. So this is limited to only manual execution in tests and some custom code. 

I am a bit on the fence with this one.

**Reply by ashb** [reply]
> The problem with setting it as Optional would mean that the mutlitude of operators (the custom ones) would have to be prepared to accept 'execute' without context, which is not backwards-incompatible I am afraid

Doesn't have to be. If we want we can inspect the function signature to check for number of args. (Not saying we _want_ to, just that it's possible to do this change without breaking existing operators.)

**Reply by dstandish** [reply]
when would you need to do any inspecting of signature?  maybe you would inspect to see if there is anything in context and act accordingly but shouldn't need to look at signature, at least within `execute`?

suppose that there's a pr changing one line, making it `def execute(self, context=None):` in base operator.  are there negative consequences of that?  the signature is changed such that context is optional.  but in some operators, they might expect it to not be empty.

if this change were adopted, would it be alright if some subclasses had `def execute(self, context)`?  pycharm would complain about inconsistent signature of subclas.  but maybe there's a way to avoid that.

**Reply by potiuk** [reply]
I think indeed context = None can work in the BaseOperator. It will only case to work if someone manually runs it without context, but in Airflow itself, we can make sure we always pass context. I think my fence leans towards changing `context=None`  in BaseOperator.

**Reply by turbaszek** [reply]
How will `context=None` work with `self.xcom_push` which requires it to be a dict with at least `ti` key? If we want to change something it will have to cover every case, including possible xcom pushes in `execute`.

**Reply by dstandish** [reply]
if `xcom_push` is called from execute, than that `execute`, would there be any problem with that operator having `def execute(self, context):`?

so that it could be forced when it is required, but not otherwise?

**Reply by turbaszek** [reply]
I'm not sure I fully understand. My point was that if we want to have `None` as default value then we have to do more changes than just setting is as default. Using `{}` as default is not recommended practice. 

Additionally I think this change should also cover `task` decorated function so they can be executed out of DAG context. And that rises similar issue as `xcom_push` but with usage of `get_current_context`.

While I like the idea of being able to execute task outside the Airflow context, I'm not sure how feasible it will be. Should we rise this discussion on dev list mailing list to gather wider feedback?

**Reply by turbaszek** [reply]
And btw. doing `context=None` in BaseOperator will not propagate to other ops where we override the `execute` method, or am I wrong?

**Reply by dstandish** [reply]
i am just asking the question ... sortof in a vacuum.... from a python perspective essentially...

consider:

```python
class BaseOperator:
    def execute(self, context=None):
        pass
```

```python
class MyXcomOp(BaseOperator):
    def execute(self, context):
        # uses context
```

Do we have a fundamental problem with doing this, aside from airflow-specific technical considerations?

**Reply by dstandish** [reply]
> context=None in BaseOperator will not propagate to other ops where we override the execute

correct but my wish is not that every operator must have context=None
i just mean, with this change, to allow it 
just, blueskying

**Reply by turbaszek** [reply]
@dstandish so I'm lost, I assume that you can just do `context=None` in your custom operator. I think this is a best way as it's operator of your design and you know that it does not use context for example (no issue with xcom_pushes for instance).

**Reply by potiuk** [reply]
> @dstandish so I'm lost, I assume that you can just do `context=None` in your custom operator. I think this is a best way as it's operator of your design and you know that it does not use context for example (no issue with xcom_pushes for instance).

Yep. You are totally right indeed Tomek. But I think what @dstandish is about is to make it more of a 'standard' so that gradually all operators will also be written in the way to accept context = None. More of a 'best practice' of what people do when they look at or copy from other operators.

**Comment by dstandish** [comment]
> you can just do context=None in your custom operator

right

but what if someone wanted to contribute an operator in airflow where context is _never_ used inside `execute`. what is wrong with having the signature be `context=None`?

that is the question here.

**Reply by turbaszek** [reply]
>  what is wrong with having the signature be context=None?

The context will always be provided and will never be `None` when `execute` is called by Airflow. So I think current signature is proper.

On the other note, changing signature in BaseOperator may cause mypy/pylint to rise issue that `execute` signature in other ops differs from parent one 😄

**Reply by potiuk** [reply]
I think there is nothing wrong with that. You can do it even now, but I think it might be considered 'bad practice'. So we are really here about changing the practice and makig it "OK"

**Reply by turbaszek** [reply]
To me this change is more like "allow airflow operators(=execute functions) to be called outside Airflow" and this is something bigger than signatures and practices. But I can be mistaken.

**Reply by potiuk** [reply]
> On the other note, changing signature in BaseOperator may cause mypy/pylint to rise issue that `execute` signature in other ops differs from parent one 😄

I think mypy/pylint will not complain now. But even if they do it can be easily changed I think.

**Reply by dstandish** [reply]
i think i had this in bash hook refactor, and it got through pre-commit

pycharm wrinkles an eyebrow (and so did @ashb ;) , hence this discussion) , but that's all

**Reply by turbaszek** [reply]
> pycharm wrinkles an eyebrow (and so did @ashb ;) , hence this discussion) , but that's all

It will also highlight any usages of `xcom_push` because it expects `Dict` not `Optional[Dict]` and here is pylint for doing `context=None` in base op:
```

************* Module airflow.operators.python
airflow/operators/python.py:111:4: W0222: Signature differs from overridden 'execute' method (signature-differs)
airflow/operators/python.py:232:4: W0222: Signature differs from overridden 'execute' method (signature-differs)
```

**Reply by dstandish** [reply]
xcom_push is not called from execute 🤔

**Reply by turbaszek** [reply]
It's quite common to call it to make it easier to access some values like this:
https://github.com/apache/airflow/blob/e7dbed2ae701bd0eff228ac0dc4787c36f885e32/airflow/providers/google/cloud/operators/automl.py#L120

So I would say it's a known practice that we have to have in mind.

**Reply by dstandish** [reply]
yeah this makes sense

in this case, i think that it would be fine for this operator to require context (even though, it possibly becomes inconsistent with parent class signature).  or if appropriate, could make it optional and handle the case appropriately in any usage of context.

**Reply by turbaszek** [reply]
As I said previously:
> To me this change is more like "allow airflow operators(=execute functions) to be called outside Airflow" and this is something bigger than signatures and practices.

So I would suggest bringing this up to the dev list.
